File: src/common/throttle.cpp
/*
 *  Copyright (c) 2020 NetEase Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
 * Project: curve
 * Created Date: 20210127
 * Author: wuhanqing
 */

#include "src/common/throttle.h"

#include <glog/logging.h>

#include <ostream>
#include <string>
#include <utility>

namespace curve {
namespace common {

const char* ThrottleTypeToName(Throttle::Type type) {
    switch (type) {
        case Throttle::Type::IOPS_TOTAL:
            return "IOPS_TOTAL";
        case Throttle::Type::IOPS_READ:
            return "IOPS_READ";
        case Throttle::Type::IOPS_WRITE:
            return "IOPS_WRITE";
        case Throttle::Type::BPS_TOTAL:
            return "BPS_TOTAL";
        case Throttle::Type::BPS_READ:
            return "BPS_READ";
        case Throttle::Type::BPS_WRITE:
            return "BPS_WRITE";
        default:
            return "Unknown";
    }
}

const std::vector<Throttle::Type> kDefaultEnabledThrottleTypes = {
    Throttle::Type::IOPS_TOTAL, Throttle::Type::IOPS_READ,
    Throttle::Type::IOPS_WRITE, Throttle::Type::BPS_TOTAL,
    Throttle::Type::BPS_READ,   Throttle::Type::BPS_WRITE};

Throttle::Throttle() : throttleParams_(), throttles_() {
    for (auto type : kDefaultEnabledThrottleTypes) {
        throttles_.emplace_back(
            type, false,
            new common::LeakyBucket(ThrottleTypeToName(type)));
    }
}

void Throttle::Stop() { throttles_.clear(); }

void Throttle::Add(bool isReadOp, uint64_t length) {
    for (auto& throttle : throttles_) {
        if (!throttle.enabled) {
            continue;
        }

        auto tokens = CalcTokens(isReadOp, length, throttle.type);
        if (tokens > 0) {
            throttle.leakyBucket->Add(tokens);
        }
    }
}

void Throttle::ResetThrottleParams(Type type, uint64_t limit, uint64_t burst,
                                   uint64_t burstLength) {
    for (auto& throttle : throttles_) {
        if (throttle.type != type) {
            continue;
        }

        throttle.enabled = (limit != 0);
        throttle.leakyBucket->SetLimit(limit, burst, burstLength);
    }
}

void Throttle::UpdateThrottleParams(const ReadWriteThrottleParams& params) {
    UpdateIfNotEqual(Type::IOPS_TOTAL, throttleParams_.iopsTotal,
                     params.iopsTotal);
    UpdateIfNotEqual(Type::IOPS_READ, throttleParams_.iopsRead,
                     params.iopsRead);
    UpdateIfNotEqual(Type::IOPS_WRITE, throttleParams_.iopsWrite,
                     params.iopsWrite);
    UpdateIfNotEqual(Type::BPS_TOTAL, throttleParams_.bpsTotal,
                     params.bpsTotal);
    UpdateIfNotEqual(Type::BPS_READ, throttleParams_.bpsRead, params.bpsRead);
    UpdateIfNotEqual(Type::BPS_WRITE, throttleParams_.bpsWrite,
                     params.bpsWrite);

    throttleParams_ = params;
}

bool Throttle::IsThrottleEnabled(Type type) const {
    for (auto& throttle : throttles_) {
        if (type != throttle.type) {
            continue;
        }

        return throttle.enabled;
    }

    return false;
}

void Throttle::UpdateIfNotEqual(
    Type type, const curve::common::ThrottleParams& oldParams,
    const curve::common::ThrottleParams& newParams) {
    if (oldParams == newParams) {
        return;
    }

    LOG(INFO) << "Update " << ThrottleTypeToName(type) << " to " << newParams;
    ResetThrottleParams(type, newParams.limit, newParams.burst,
                        newParams.burstSeconds);
}

inline bool Throttle::IsWriteThrottle(Type type) const {
    return type == Type::IOPS_WRITE || type == Type::BPS_WRITE;
}

inline bool Throttle::IsReadThrottle(Type type) const {
    return type == Type::IOPS_READ || type == Type::BPS_READ;
}

inline bool Throttle::IsIOPSThrottle(Type type) const {
    return type == Type::IOPS_TOTAL || type == Type::IOPS_READ ||
           type == Type::IOPS_WRITE;
}

uint64_t Throttle::CalcTokens(bool isRead, uint64_t length, Type type) const {
    if (isRead && IsWriteThrottle(type)) {
        return 0;
    } else if (!isRead && IsReadThrottle(type)) {
        return 0;
    }

    return IsIOPSThrottle(type) ? 1 : length;
}

}  // namespace common
}  // namespace curve

File: tools-v2/pkg/cli/command/curvebs/update/throttle/throttle.go
/*
*  Copyright (c) 2023 NetEase Inc.
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
 */
/*
* Project: curve
* Created Date: 2023-04-24
* Author: chengyi01
 */

package throttle

import (
	"context"

	cmderror "github.com/opencurve/curve/tools-v2/internal/error"
	cobrautil "github.com/opencurve/curve/tools-v2/internal/utils"
	basecmd "github.com/opencurve/curve/tools-v2/pkg/cli/command"
	"github.com/opencurve/curve/tools-v2/pkg/config"
	"github.com/opencurve/curve/tools-v2/pkg/output"
	"github.com/opencurve/curve/tools-v2/proto/proto/nameserver2"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"google.golang.org/grpc"
)

type UpdateFileThrottleRpc struct {
	Info      *basecmd.Rpc
	Request   *nameserver2.UpdateFileThrottleParamsRequest
	mdsClient nameserver2.CurveFSServiceClient
}

var _ basecmd.RpcFunc = (*UpdateFileThrottleRpc)(nil) // check interface

func (uRpc *UpdateFileThrottleRpc) NewRpcClient(cc grpc.ClientConnInterface) {
	uRpc.mdsClient = nameserver2.NewCurveFSServiceClient(cc)
}

func (uRpc *UpdateFileThrottleRpc) Stub_Func(ctx context.Context) (interface{}, error) {
	return uRpc.mdsClient.UpdateFileThrottleParams(ctx, uRpc.Request)
}

type ThrottleCommand struct {
	basecmd.FinalCurveCmd
	Rpc      *UpdateFileThrottleRpc
	Response *nameserver2.UpdateFileThrottleParamsResponse
}

var _ basecmd.FinalCurveCmdFunc = (*ThrottleCommand)(nil) // check interface

func NewUpdateThrottleCommand() *ThrottleCommand {
	throttleCmd := &ThrottleCommand{
		FinalCurveCmd: basecmd.FinalCurveCmd{
			Use:     "throttle",
			Short:   "update file throttle params",
		},
	}
	basecmd.NewFinalCurveCli(&throttleCmd.FinalCurveCmd, throttleCmd)
	return throttleCmd
}

func NewThrottleCommand() *cobra.Command {
	return NewUpdateThrottleCommand().Cmd
}

func (tCmd *ThrottleCommand) AddFlags() {
	config.AddRpcRetryTimesFlag(tCmd.Cmd)
	config.AddRpcTimeoutFlag(tCmd.Cmd)
	config.AddBsMdsFlagOption(tCmd.Cmd)
	config.AddBsPathRequiredFlag(tCmd.Cmd)
	config.AddBsThrottleTypeRequiredFlag(tCmd.Cmd)
	config.AddBsUserOptionFlag(tCmd.Cmd)
	config.AddBsPasswordOptionFlag(tCmd.Cmd)
	config.AddBsLimitRequiredFlag(tCmd.Cmd)
	config.AddBsBurstOptionFlag(tCmd.Cmd)
	config.AddBsBurstLengthOptionFlag(tCmd.Cmd)
}

func (tCmd *ThrottleCommand) Init(cmd *cobra.Command, args []string) error {
	path := config.GetBsFlagString(cmd, config.CURVEBS_PATH)
	throttleTypeStr := config.GetBsFlagString(cmd, config.CURVEBS_TYPE)
	throttleType, err := cobrautil.ParseThrottleType(throttleTypeStr)
	if err.TypeCode() != cmderror.CODE_SUCCESS {
		return err.ToError()
	}
	limit := config.GetBsFlagUint64(cmd, config.CURVEBS_LIMIT)
	params := &nameserver2.ThrottleParams{
		Type:  &throttleType,
		Limit: &limit,
	}
	if config.GetFlagChanged(cmd, config.CURVEBS_BURST) {
		burst := config.GetBsFlagUint64(cmd, config.CURVEBS_BURST)
		burstLength := config.GetBsFlagUint64(cmd, config.CURVEBS_BURST_LENGTH)
		if burstLength == 0 {
			burstLength = 1
		}
		params.Burst = &burst
		params.BurstLength = &burstLength
	}
	date, errDat := cobrautil.GetTimeofDayUs()
	owner := config.GetBsFlagString(tCmd.Cmd, config.CURVEBS_USER)
	if errDat.TypeCode() != cmderror.CODE_SUCCESS {
		return errDat.ToError()
	}
	request := &nameserver2.UpdateFileThrottleParamsRequest{
		FileName:       &path,
		Owner:          &owner,
		Date:           &date,
		ThrottleParams: params,
	}
	password := config.GetBsFlagString(tCmd.Cmd, config.CURVEBS_PASSWORD)
	if owner == viper.GetString(config.VIPER_CURVEBS_USER) && len(password) != 0 {
		strSig := cobrautil.GetString2Signature(date, owner)
		sig := cobrautil.CalcString2Signature(strSig, password)
		request.Signature = &sig
	}
	mdsAddrs, errMds := config.GetBsMdsAddrSlice(tCmd.Cmd)
	if errMds.TypeCode() != cmderror.CODE_SUCCESS {
		return errMds.ToError()
	}
	timeout := config.GetFlagDuration(tCmd.Cmd, config.RPCTIMEOUT)
	retrytimes := config.GetFlagInt32(tCmd.Cmd, config.RPCRETRYTIMES)
	tCmd.Rpc = &UpdateFileThrottleRpc{
		Info:    basecmd.NewRpc(mdsAddrs, timeout, retrytimes, "UpdateFileThrottle"),
		Request: request,
	}
	tCmd.SetHeader([]string{cobrautil.ROW_RESULT})
	return nil
}

func (tCmd *ThrottleCommand) Print(cmd *cobra.Command, args []string) error {
	return output.FinalCmdOutput(&tCmd.FinalCurveCmd, tCmd)
}

func (tCmd *ThrottleCommand) RunCommand(cmd *cobra.Command, args []string) error {
	result, err := basecmd.GetRpcResponse(tCmd.Rpc.Info, tCmd.Rpc)
	if err.TypeCode() != cmderror.CODE_SUCCESS {
		return err.ToError()
	}
	tCmd.Response = result.(*nameserver2.UpdateFileThrottleParamsResponse)
	tCmd.Result = tCmd.Response
	tCmd.Error = cmderror.ErrUpdateFileThrottle(tCmd.Response.GetStatusCode(), tCmd.Rpc.Request.GetFileName())
	tCmd.TableNew.Append([]string{tCmd.Error.Message})
	if tCmd.Error.TypeCode() != cmderror.CODE_SUCCESS {
		return tCmd.Error.ToError()
	}
	return nil
}

func (tCmd *ThrottleCommand) ResultPlainOutput() error {
	return output.FinalCmdOutputPlain(&tCmd.FinalCurveCmd)
}
